File: ios-bookmind/Sources/BookMemory/HeuristicCharacterDetector.swift
import Foundation
import SharedModels

/// Detects characters by extracting proper nouns / capitalized name-like tokens.
/// Naive but adequate as a first pass; replace with NER later.
public final class HeuristicCharacterDetector: CharacterDetecting {
    public init() {}

    private static let stopWords: Set<String> = [
        "The", "A", "An", "He", "She", "It", "They", "We", "I", "You",
        "Chapter", "Part", "Book", "Mr", "Mrs", "Ms", "Dr",
        "Глава", "Часть", "Книга", "Господин", "Госпожа"
    ]

    public func detectCharacters(
        in chapter: Chapter,
        rawText: String,
        existing: [Character]
    ) -> [Character] {
        let names = extractCandidateNames(from: rawText)
        var byCanonical: [String: Character] = [:]
        for c in existing { byCanonical[c.canonicalName.lowercased()] = c }

        var result: [Character] = []
        for name in names {
            let key = name.lowercased()
            if let prior = byCanonical[key] {
                // Extend last safe chapter index up to current chapter.
                if chapter.index > prior.lastSafeChapterIndex {
                    let updated = Character(
                        id: prior.id,
                        bookID: prior.bookID,
                        canonicalName: prior.canonicalName,
                        aliases: prior.aliases,
                        firstChapterIndex: prior.firstChapterIndex,
                        lastSafeChapterIndex: chapter.index,
                        safeSummary: prior.safeSummary
                    )
                    byCanonical[key] = updated
                    result.append(updated)
                }
            } else {
                let new = Character(
                    id: "\(chapter.bookID.rawValue)#char-\(stableSlug(name))",
                    bookID: chapter.bookID,
                    canonicalName: name,
                    aliases: [],
                    firstChapterIndex: chapter.index,
                    lastSafeChapterIndex: chapter.index,
                    safeSummary: nil
                )
                byCanonical[key] = new
                result.append(new)
            }
        }
        return result
    }

    func extractCandidateNames(from text: String) -> [String] {
        // Words starting with an uppercase letter, appearing >=2 times, length 2-30.
        let pattern = #"\b\p{Lu}\p{L}{1,29}\b"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }
        let ns = text as NSString
        let matches = regex.matches(in: text, range: NSRange(location: 0, length: ns.length))

        var freq: [String: Int] = [:]
        for m in matches {
            let token = ns.substring(with: m.range)
            if Self.stopWords.contains(token) { continue }
            freq[token, default: 0] += 1
        }
        // Filter: appears at least twice OR clearly a name structure (length >= 3).
        return freq
            .filter { $0.value >= 2 && $0.key.count >= 3 }
            .map(\.key)
            .sorted()
    }

    private func stableSlug(_ s: String) -> String {
        s.lowercased()
            .replacingOccurrences(of: #"[^\p{L}\p{N}]+"#, with: "-", options: .regularExpression)
            .trimmingCharacters(in: CharacterSet(charactersIn: "-"))
    }
}

File: android-bookmind/app/src/main/java/com/bookmind/memory/HeuristicCharacterDetector.kt
package com.bookmind.memory

import com.bookmind.core.model.Chapter
import com.bookmind.core.model.Character
import javax.inject.Inject

/**
 * = iOS `HeuristicCharacterDetector`.
 * Detects characters by extracting capitalized, repeated name-like tokens.
 */
class HeuristicCharacterDetector @Inject constructor() : CharacterDetecting {

    override fun detectCharacters(
        chapter: Chapter,
        rawText: String,
        existing: List<Character>
    ): List<Character> {
        val names = extractCandidateNames(rawText)
        val byCanonical = HashMap<String, Character>()
        for (c in existing) byCanonical[c.canonicalName.lowercase()] = c

        val result = mutableListOf<Character>()
        for (name in names) {
            val key = name.lowercase()
            val prior = byCanonical[key]
            if (prior != null) {
                if (chapter.index > prior.lastSafeChapterIndex) {
                    val updated = prior.copy(lastSafeChapterIndex = chapter.index)
                    byCanonical[key] = updated
                    result.add(updated)
                }
            } else {
                val new = Character(
                    id = "${chapter.bookID.raw}#char-${stableSlug(name)}",
                    bookID = chapter.bookID,
                    canonicalName = name,
                    aliases = emptyList(),
                    firstChapterIndex = chapter.index,
                    lastSafeChapterIndex = chapter.index,
                    safeSummary = null
                )
                byCanonical[key] = new
                result.add(new)
            }
        }
        return result
    }

    /** Words starting with an uppercase letter, appearing >= 2 times, length 3-30. */
    fun extractCandidateNames(text: String): List<String> {
        val freq = HashMap<String, Int>()
        for (match in namePattern.findAll(text)) {
            val token = match.value
            if (token in stopWords) continue
            freq[token] = (freq[token] ?: 0) + 1
        }
        return freq
            .filter { it.value >= 2 && it.key.length >= 3 }
            .keys
            .sorted()
    }

    private fun stableSlug(s: String): String =
        s.lowercase()
            .replace(Regex("""[^\p{L}\p{N}]+"""), "-")
            .trim('-')

    companion object {
        private val namePattern = Regex("""\b\p{Lu}\p{L}{1,29}\b""")

        private val stopWords: Set<String> = setOf(
            "The", "A", "An", "He", "She", "It", "They", "We", "I", "You",
            "Chapter", "Part", "Book", "Mr", "Mrs", "Ms", "Dr",
            "Глава", "Часть", "Книга", "Господин", "Госпожа"
        )
    }
}
